homeassistant/components/tplink_lte.py

```diff
@@ -0,0 +1,125 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = 'tplink_lte'
+DATA_KEY = 'tplink_lte'
+
+DEFAULT_DISCOVERY = True
+
+_NOTIFY_SCHEMA = vol.All(vol.Schema({
+    vol.Optional(CONF_NAME): cv.string,
+    vol.Required(ATTR_TARGET): vol.All(cv.ensure_list, [cv.string]),
+}))
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.All(cv.ensure_list, [vol.Schema({
+        vol.Required(CONF_HOST): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Optional(SERVICE_NOTIFY):
+            vol.All(cv.ensure_list, [_NOTIFY_SCHEMA]),
+    })])
+}, extra=vol.ALLOW_EXTRA)
+
+
+@attr.s
+class ModemData:
+    """Class for modem state."""
+
+    modem = attr.ib()
+
+
+@attr.s
+class LTEData:
+    """Shared state."""
+
+    websession = attr.ib()
+    modem_data = attr.ib(init=False, factory=dict)
+
+    def get_modem_data(self, config):
+        """Get the requested or the only modem_data value."""
+        if CONF_HOST in config:
+            return self.modem_data.get(config[CONF_HOST])
+        if len(self.modem_data) == 1:
+            return next(iter(self.modem_data.values()))
+
+        return None
+
+
+async def async_setup(hass, config):
+    """Set up TP-Link LTE component."""
+    if DATA_KEY not in hass.data:
+        websession = async_create_clientsession(
+            hass, cookie_jar=aiohttp.CookieJar(unsafe=True))
+        hass.data[DATA_KEY] = LTEData(websession)
+
+    domain_config = config.get(DOMAIN, [])
+
+    tasks = [_setup_lte(hass, conf) for conf in domain_config]
+    if tasks:
+        await asyncio.wait(tasks)
+
+    for conf in domain_config:
+        for notify_conf in conf.get(SERVICE_NOTIFY, []):
+            hass.async_create_task(discovery.async_load_platform(
+                hass, 'notify', DOMAIN, notify_conf, config))
+
+    return True
+
+
+async def _setup_lte(hass, lte_config, delay=0):
+    """Set up a TP-Link LTE modem."""
+    import tp_connected
+
+    setup_task = None
+    try:
+        if delay:
+            await asyncio.sleep(delay)
+    except asyncio.CancelledError:
+        return
+
+    host = lte_config[CONF_HOST]
+    password = lte_config[CONF_PASSWORD]
+
+    websession = hass.data[DATA_KEY].websession
+    modem = tp_connected.Modem(hostname=host, websession=websession)
+
+    try:
+        await modem.login(password=password)
+    except asyncio.CancelledError:
+        return
+    except tp_connected.Error:
+        delay = max(15, min(2*delay, 300))
+        _LOGGER.warning("Retrying %s in %d seconds", host, delay)
+        setup_task = hass.loop.create_task(_setup_lte(hass, lte_config, delay))
+        return
+
+    modem_data = ModemData(modem)
+    hass.data[DATA_KEY].modem_data[host] = modem_data
+
+    async def cleanup(event):
+        """Clean up resources."""
+        if setup_task is not None and not setup_task.done():
+            setup_task.cancel()
+        await modem.logout()
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, cleanup)
```

Comment:
I think we need to move this up after creating the retry setup task and before returning. Otherwise we will never register the listener until we've succeeded with setup.

We should also remove the existing listener, by calling the return value of the registration, each time before registering a new one.

Reply:
FYI, I added the task cleanup to netgear_lte now: #18163

My proposal avoids removal of the existing listener by running a loop. I found that simpler to get working.

Reply:
I like that!

Reply:
@amelchio thanks for the hint, I reused your solution with the addition of two `CancelledError` exception handlers, hope that are useful there

Reply:
I believe catching `CancelledError` is only necessary if we want to do something, like release resources.

Reply:
As @MartinHjelmare pointed out in a older commit with a `CancelledError` handler we avoid to print a not useful stack trace in some cases. Does homeassistant loop handle this for us?

Reply:
Sorry, I think @amelchio  is correct. Eg:
```py
In [22]: import asyncio

In [23]: async def running_task():
    ...:     while True:
    ...:         print('running')
    ...:         print('sleeping')
    ...:         await asyncio.sleep(5)
    ...:         

In [24]: def task_canceller(task):
    ...:     print('in task_canceller')
    ...:     task.cancel()
    ...:     print('canceled the task')
    ...:     
    ...:     

In [25]: async def main(loop):
    ...:     print('creating task')
    ...:     task = loop.create_task(running_task())
    ...:     loop.call_soon(task_canceller, task)
    ...:     print('sleeping 2 secs in main')
    ...:     await asyncio.sleep(2)
    ...:     print('finished main')
    ...:     
    ...:     

In [26]: loop = asyncio.new_event_loop()

In [27]: try:
    ...:     loop.run_until_complete(main(loop))
    ...: finally:
    ...:     loop.close()
    ...:     
creating task
sleeping 2 secs in main
running
sleeping
in task_canceller
canceled the task
finished main
```